### PR TITLE
feat(babel-plugin-component): remove import dedupe

### LINKED ISSUE #5023: [babel-plugin-component] Remove `import` dedupe transformation
Opening this issue for posterity; in the breaking changes wishlist (#2964) we said:

- remove [dedupe imports transformation](https://github.com/salesforce/lwc/blob/4b8b03e0b4246205984a020c4fb59e92a18ad4ac/packages/%40lwc/babel-plugin-component/src/index.js#L51-L52) from Babel plugin component

The benefit here is that we could delegate to other parts of the compiler stack (e.g. Rollup) to handle duplicate `import`s. This would be code cleanup mostly.

### PR DESCRIPTION
**Pull request by wjhsf**
## Details

`@lwc/babel-plugin-component` does import deduping, but this is unnecessary as it is also done by rollup. This is a potentially breaking change if code relies on the order of imports. However, it only impacts direct consumers of `@lwc/babel-plugin-component` and `@lwc/compiler`. It does not impact consumers of `@lwc/rollup-plugin` or users of a bundler (like rollup) that does deduping.

Closes #5023.

## Does this pull request introduce a breaking change?

- 😮‍💨 No, it does not introduce a breaking change.
- 💔 Yes, it does introduce a breaking change.

## Does this pull request introduce an observable change?

- 🤞 No, it does not introduce an observable change.
- 🔬 Yes, it does include an observable change.

### CODE REVIEW

File: packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// @lwc/babel-plugin-component prior to v9 and we want to ensure there are no regressions
+it('merges duplicate imports', async () => {
+    const { code, warnings } = await runRollup('fixtures/basic/basic.js', {});
+    const lwcImport = /import \{ ([^}]+?) \} from 'lwc'/g;
+    const imports = Array.from(code.matchAll(lwcImport));
+    expect(warnings).toHaveLength(0);
+    expect(imports).toHaveLength(1);
+    const specifiers = imports[0][1];
+    expect(specifiers).toContain('renderer');
+    expect(specifiers).not.toContain('renderer as');
```
Comment (jye-sf) [inline packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts:51]: What is the expected output of the final result for `basic.js` here?

**Review by jye-sf — CHANGES_REQUESTED** [review]
The description mentions that direct consumers of @lwc/compiler will also be impacted by this change. How will that be affected, and can we get tests for that?

File: packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts (hunk 1 above)
Comment (wjhsf) [reply packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts:51]: The relevant output is `import { LightningElement, renderer } from 'lwc'`. The multiple imports in `basic.js` should be merged and they should not use `as`. We don't check for that exact string because there are other imports added as part of compilation (`registerComponent`) and we don't care about the ordering.

File: packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// This test validates the behavior of rollup because import deduping was handled by
+// @lwc/babel-plugin-component prior to v9 and we want to ensure there are no regressions
+it('merges duplicate imports', async () => {
+    const { code, warnings } = await runRollup('fixtures/basic/basic.js', {});
+    const lwcImport = /import \{ ([^}]+?) \} from 'lwc'/g;
+    const imports = Array.from(code.matchAll(lwcImport));
+    expect(warnings).toHaveLength(0);
+    expect(imports).toHaveLength(1);
+    const specifiers = imports[0][1];
+    expect(specifiers).toContain('renderer');
```
Comment (wjhsf) [inline packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts:50]:
```suggestion
    // `specifiers` should include `LightningElement` and `renderer` from basic.js
    // we don't do an exact string match because the compiler adds imports
    expect(specifiers).toContain('renderer');
```

File: packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts (hunk 1 above)
Comment (wjhsf) [reply packages/@lwc/rollup-plugin/src/__tests__/importDedupe/importDedupe.spec.ts:51]: added a code comment to clarify

**Comment by wjhsf** [comment]
> The description mentions that direct consumers of @lwc/compiler will also be impacted by this change. How will that be affected, and can we get tests for that?

The actual change is in `@lwc/babel-plugin-component`. The tests in that package have been updated with the new expected output. `@lwc/compiler` is just an extra wrapper on top of that.

```diff
- import { registerDecorators as _registerDecorators, LightningElement, registerComponent as _registerComponent } from "lwc";
+ import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./test.html";
+ import { LightningElement, registerComponent as _registerComponent } from 'lwc';
```

This is the change. After using just the babel plugin, there are two imports from `lwc`, rather than one. And, importantly, some are _after_ `./test.html`, rather than all being before. If your code relies on import order (to set globals or something), then this is potentially breaking. (Weird edge case, but plausible.)

**Comment by jye-sf** [comment]
@wjhsf For some reason I thought we used rollup in @lwc/compiler. My bad.

File: packages/@lwc/babel-plugin-component/src/__tests__/fixtures/api-decorator/@api-setter-on-one-class-member-should-not-conflict-with-@api-getter-on-another/expected.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
-import { registerDecorators as _registerDecorators, LightningElement, registerComponent as _registerComponent } from "lwc";
+import { registerDecorators as _registerDecorators } from "lwc";
 import _tmpl from "./test.html";
+import { LightningElement, registerComponent as _registerComponent } from 'lwc';
```
Comment (jye-sf) [inline packages/@lwc/babel-plugin-component/src/__tests__/fixtures/api-decorator/@api-setter-on-one-class-member-should-not-conflict-with-@api-getter-on-another/expected.js:3]: Since we're changing the public API of @lwc/compiler, we definitely need some sanity tests for this use case.

**Comment by jye-sf** [comment]
> However, it only impacts direct consumers of @lwc/babel-plugin-component and @lwc/compiler.
I think a big use case to verify will be LWR, which uses `@lwc/compiler` directly. Otherwise AFAIK the platform compiler is the other major use case.

**Comment by wjhsf** [comment]
I took a deeper look at the compiler, and here are the relevant transformations that it does:
1. If it's transforming a component file (`export default class extends LightningElement {}`), then it inserts a template import (`import _html from "./test.html"`) at the top of the file.
2. If a decorator is used, it adds `import { registerDecorators as _registerDecoratores } from 'lwc'` above `import _tmpl`.
3. If a component feature flag is used, it adds the feature flag import (`import featureFlag from '@salesforce/featureFlag/FLAG_NAME`) above `import _tmpl`. 
    a. This is a new feature and the imported module is most likely unique, so I don't think we need to consider it for dedupe breakage. 
4. Previously, all `import` statements were merged into the first occurrence of each imported module. Now they are left as written. 

```js
// actual.js
import { bar } from "foo";
globalThis.modified = true;
import { qux } from "foo";

// expected.js before
import { bar, qux } from "foo";
globalThis.modified = true;

// expected.js after
import { bar } from "foo";
globalThis.modified = true;
import { qux } from "foo";
```

Note that the order of imports only matters if the exports rely on a global state _and_ it's loaded in a non-standard way. Both ESM `import` and CJS `require` leverage caching, so that the imported module is only evaluated once.

~🤔 All synthetic shadow tests are currently failing. That's sus...~ nvm that's also happening in other PRs
